src/napari/_qt/layer_controls/widgets/qt_contrast_limits.py

```diff
@@ -0,0 +1,228 @@
+
+    def mousePressEvent(self, event):
+        """Update the slider, or, on right-click, pop-up an expanded slider.
+
+        The expanded slider provides finer control, directly editable values,
+        and the ability to change the available range of the sliders.
+
+        Parameters
+        ----------
+        event : napari.utils.event.Event
+            The napari event that triggered this method.
+        """
+        if event.button() == Qt.MouseButton.RightButton:
+            self.show_clim_popup.emit()
+        else:
+            super().mousePressEvent(event)
+
+
+class QContrastLimitsPopup(QRangeSliderPopup):
+    def __init__(self, layer: 'Image', parent=None) -> None:
+        super().__init__(parent)
+
+        decimals = range_to_decimals(layer.contrast_limits_range, layer.dtype)
+        self.slider.setRange(*layer.contrast_limits_range)
+        self.slider.setDecimals(decimals)
+        self.slider.setSingleStep(10**-decimals)
+        self.slider.setValue(layer.contrast_limits)
+
+        connect_setattr(self.slider.valueChanged, layer, 'contrast_limits')
+        connect_setattr(
+            self.slider.rangeChanged, layer, 'contrast_limits_range'
+        )
+
+        def reset():
+            layer.reset_contrast_limits()
+            layer.contrast_limits_range = layer.contrast_limits
+            decimals_ = range_to_decimals(
+                layer.contrast_limits_range, layer.dtype
+            )
+            self.slider.setDecimals(decimals_)
+            self.slider.setSingleStep(10**-decimals_)
+
+        reset_btn = QPushButton('reset')
+        reset_btn.setObjectName('reset_clims_button')
+        reset_btn.setToolTip(trans._('Autoscale contrast to data range'))
+        reset_btn.setFixedWidth(45)
+        reset_btn.clicked.connect(reset)
+        self._layout.addWidget(
+            reset_btn, alignment=Qt.AlignmentFlag.AlignBottom
+        )
+
+        # the "full range" button doesn't do anything if it's not an
+        # unsigned integer type (it's unclear what range should be set)
+        # so we don't show create it at all.
+        if np.issubdtype(normalize_dtype(layer.dtype), np.integer):
+            range_btn = QPushButton('full range')
+            range_btn.setObjectName('full_clim_range_button')
+            range_btn.setToolTip(
+                trans._('Set contrast range to full bit-depth')
+            )
+            range_btn.setFixedWidth(75)
+            range_btn.clicked.connect(layer.reset_contrast_limits_range)
+            self._layout.addWidget(
+                range_btn, alignment=Qt.AlignmentFlag.AlignBottom
+            )
+
+
+class QtContrastLimitsSliderControl(QtWidgetControlsBase):
+    """
+    Class that wraps the connection of events/signals between the layer contrast
+    limits and Qt widgets.
+
+    Parameters
+    ----------
+    parent: qtpy.QtWidgets.QWidget
+        An instance of QWidget that will be used as widgets parent
+    layer : napari.layers.Layer
```

Comment:
Again `Image` or `Surface` ?

Reply:
I think both as previously suggested! So:

```suggestion
    layer : napari.layers.Image | napari.layers.Surface
```